### ISSUE TITLE
Match event loop wakeups to expected messages

### ISSUE THREAD

**Issue by jdm** [servo/servo#23865]
@gterzian left a series of interesting comments about the fragility of the current system that wakes up the main thread to process events in https://github.com/servo/servo/issues/23832#issuecomment-515323314. I agree that it would be good to find a system that would better express the intention of waking up the event loop so that it doesn't go back to sleep until all expected events have been processed. One idea I'm playing with looks like this:
* the main thread stores a mapping of different kinds of integrated event loops (webgl, webxr, compositing) to number of expected events
* when a consumer posts an event, the event loop waker interface is used to register a new event of a particular type is expected
* when this wakeup is processed, the count of expected events is increased by one
* the main thread uses try_recv when processing receivers; for each event processed it decreases the appropriate expected count by one
* if there are non-zero expected messages of any type, the embedder is notified that more processing is required
* the embedder is responsible for not returning to a quiescent state until servo reaches zero expected messages

Possible difficulties:
* if the main thread is eagerly processing messages, how do we tell if a wake up event should actually increase the count of expected messages? if we end up "waiting" for an event that's already been processed, that will lead to unnecessary CPU usage.

**Comment by jdm** [comment]
One conceptually nice system would be to have wake up events include an increasing monotonic id that's associated with the events that need to be processed so we could easily check if we have already processed them. That _might_ be possible if all senders of the events live in the same process and could share an AtomicUint that gets included in the message.

**Comment by gterzian** [comment]
Another way to frame the problem, is that if those event-loops ran on their own thread of execution(whether a thread or some form of task/virtual thread), we wouldn't have a problem since they could just block on a `select` or `recv`, and be woken up when a message/event is ready to be handled. 

But when due to "main thread affinity" of some device system we need to run a bunch of disparate event-loops on the same "main" thread, you end-up with complicated logic using `try_rcv` or `recv_timeout` because one logically separate "event-loop/sub-system" is unable to block on incoming events/messages, because it needs to share the "turn of the event-loop" with other "event-loop/sub-system". 

The main risks are unintended consequences in the form of:

1. weird hangs or deadlocks because it's too hard to manage the potential interactions between the various "event-loop/sub-system" as a single event-loop, and you can't remain aware of underlying dependencies between events/messages. 
2. spurious wake-ups of the main-thread, since you can't be sure that the "message that woke it up" is actually available on the channel(the wake-up and the "sending a message" are not atomic or transactional in any way). Or perhaps the wake-up has no effect if the main-thread is already scheduled, yet it "misses" the new message with isn't available yet.  Spurious wake-ups could actually result in indefinite hangs, since a wake-up is "too quick", and then when the message becomes available, the main-thread doesn't wake-up again(since you're not actually blocking on a `recv`, you're trying to catch a message while you're running using various `try_recv` and `recv_timeout`).

**Comment by gterzian** [comment]
> My temptation would be to rewrite it using async/await which is exactly intended to solve problems like this. 

@asajeffrey If we were put each disparate "event-loop/sub-system" in its own task, and then run a single-threaded async/await runtime that would "wake-up" specific tasks when specific events come-in, it could certainly be doable. On the Servo side, you'd just have to switch to a task-aware channel to handle incoming messages on the ipc router, which would result in waking up a specific "task" representing a sub-system like WebXR or WebGL, all running in the same thread. 

The main difficulty would be running the embedder event-loop(the actual Hololens or ML code) as a task on that same runtime(you'd loose the thread affinity otherwise). 

It could be easier to use a workflow where the call to the "waker" includes an specific event, which is then handled by the embedder event-loop, by which I mean:

In the case of Hololens

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/support/hololens/ServoApp/BrowserPage.cpp#L136

As you can see, the current "wake up" handler already enqueues an event on that event-loop: 

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/support/hololens/ServoApp/BrowserPage.cpp#L254

So all we'de need is to make that event a specific event, and handle it as a case, like is currently done for `Event::FORWARD` at 

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/support/hololens/ServoApp/BrowserPage.cpp#L148

So a `Event::WebGL` would then call something like `mServo->HandleWebGL();`

Such a call then ultimately results in a specific `WindowEvent` being received at 

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/components/servo/lib.rs#L544

And at that point, it's a bit like you're inside a task that was woken up in an async/await context because of a specific event. 

And then you can actually block on your underlying `recv` on a channel, since you can be sure there is an incoming webgl command, for example. No need to `try_recv`, or a timeout, and so on. 

The problem with that, which I hadn't thought of in my initial sketch in the other issue, is that you wouldn't be able to drain the channel with any other available message, since doing so would then result in the next call to `recv` to block indefinitely(since you'd receiving one `Event::WebGL` per message in the underlying channel, so if you drain the channel the next time you get a `Event::WebGL` you can't be sure there is a message available). 

The workflow sketched by @jdm above is more sophisticated, and I think a first version could just have one `Event` per call to `main_threadwaker.wake(Event)`, resulting in one `WindowEvent` being received by Servo on the main thread, resulting in one message being received on an underlying channel using `recv`. 

That way you don't need to count events and so on, you just have one event per "servo operation" on the main-thread. The downside is you can't drain the channel inside one such operation, and I wonder if that really matters. Since the main thread is driven by the system in which it is embedded, and it's some sort of priority thread I suppose, it shouldn't really matter to have a one-to-one mapping of event to operation.

**Comment by gterzian** [comment]
For ML it looks like it would require the `EventLoopWakerInstance`

https://github.com/servo/servo/blob/ad82d67b2e5ab56d046274ebb5436c3388ddf21a/ports/libmlservo/src/lib.rs#L410

To call back into servo, looks like it could be done even without calling into the specific ML app code?

Or maybe here you could actually block on an incoming event from the waker?

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/support/magicleap/Servo3D/Servo3D.cpp#L245

**Comment by gterzian** [comment]
Also important:

> The current code for messages to the main thread is a bit brittle, and assumes that while an rAF is being processed, you don't need to wake up the main thread, because of https://github.com/servo/servo/blob/master/components/compositing/compositor.rs#L1010. This is the same assumption animations make: the compositor has two modes, one where animations are running (and the compositor runs once per frame) and one where they're not (and the compositor blocks on window events). This isn't great, especially at phase transitions, but cleaning it up most likely means waiting for The Big Main Thread Rewrite.

https://github.com/servo/servo/blob/master/components/compositing/compositor.rs#L1010

is called in these places:

https://github.com/servo/servo/blob/8f7440f36881fa60f4237d0dec8928799a6aa747/components/compositing/compositor.rs#L483

https://github.com/servo/servo/blob/8f7440f36881fa60f4237d0dec8928799a6aa747/components/compositing/compositor.rs#L445

https://github.com/servo/servo/blob/5c8fee4e0bdf01efc83f4db10a60ac1629db6795/components/servo/lib.rs#L548

https://github.com/servo/servo/blob/8f7440f36881fa60f4237d0dec8928799a6aa747/components/compositing/compositor.rs#L1458


Some of these calls originate from "inside" Servo, for example by calling `CompositorProxy.recomposite` or by the constellation in response to a `WebDriverCommandMsg::TakeScreenshot`, and others from the embedding device, for example when `heartbeat_servo` is called by ML. 

All could be unfied by having them go through the embedding device event-loop, which would then forward the appropriate event to main-thread Servo. 

And it could support process boundaries too. 

Let's say we want to put the constellation into it's own process, and put "main-thread Servo" into a "Device service process", then when the constellation sends an IPC message like `ToCompositorMsg::CreatePng(reply)`, it could be initially handled by a router thread inside the "Device service process", which would then call into the "wake" api(which probably needs a better name), which would result in an event on the ML/HoloLens device event-loop, which would then call back into Servo and result in a `WindowEvent` being handled by main-thread servo. 

Currently, when the constellation sends a `ToCompositorMsg::CreatePng(reply)`, it will already be handled on the big `handle_events` part of "main-thread Servo", as part of `self.compositor.receive_messages()`. 

So basically, all we need to do is break-up `handle_events` to only handle `WindowEvent`, which would all come via the event-loop of the embedding device. 

"Waking" the event-loop then becomes like sending a specific message to "main-thread Servo", with also the potential to handle it directly on the embedder event-loop(like the turn "animation mode" on and off could just be done directly in the ML/Hololens code, which would sent loop and tell servo to update one event at a time).

**Comment by gterzian** [comment]
Ok, so I just realized what we're actually talking about here is extending `embedder_traits` to cover all these new use-cases like Webgl on the main-thread and WebXr. 

We currently have a `EmbedderProxy`:

https://github.com/servo/servo/blob/f78dd6142e0b87492e0f8d5668c333a18a0cf780/components/embedder_traits/lib.rs#L69

Which is now using the `EventloopWaker` which is actually the `webxr_api::MainThreadWaker`, which actually points us in the right direction, I think. 

How come the `EmbedderProxy` uses a normal channel, not an IPC one? Because currently, when script sends a message to the embedder, it is routed via the constellation, so the first leg from script to constellation is over IPC, but the last leg constellation to embedder is not(but probably should be, if we want the constellation lifecyle to not match that of windows, devices, and so on). 

And then we have WebXR which for example does a "request animation frame" from script, and uses a `SessionMsg` from that api. 

That `SessionMsg` will then be handled, on the main-thread/process, via a call to `run_one_frame`, inside `self.compositor.perform_updates()` of the compositor, itself called in `handle_events` of "main-thread Servo" at 

https://github.com/servo/servo/blob/5c8fee4e0bdf01efc83f4db10a60ac1629db6795/components/servo/lib.rs#L731

So I think the way WebXR is messaging is optimal, and it would be even better if it went through `EmbedderProxy`, with the latter being refactored to message directly with the embedder.

---- 

So the WebXR call could really be a `EmbedderMsg::XrRaf(PipelinId)`.

Then running WebGL on the main thread could really mean: "send messages to the embedder", in fact using webgl probably could always mean that, with the embedder either running the command on the same thread or forwarding it to the dedicated thread. 

And the main change would be that the message is received on a "router thread", however the router thread would not be a normal IPC router, it would route "events"  to the main thread, which would be handled by the device event-loop. 

Then, the device event-loop would do what it needs to do, if anything, and enqueue a `WindowEvent` for "main-thread Servo" to handle. 

More concretely in case of a Xr raf:

1. The router receives `EmbedderMsg::XRRaf(pipelineId, dest)`. 
2. It writes to a piece of data shared with the main-thread:
        a. `let timestamp = now`.
        a. Get `xr_session` from `HashMap<PipelineId, XrSession>`,
        b. `xr_session.animation_frame_request_start(timestamp, dest)`;
3. It calls into `EventloopWaker.wake(Event::Raf(pipelineId)`. 
4. This calls into a device specific way to add an event to the queue, in Hololens, something like `SendEventToServo({{Event::RAF}, {pipelineId}});`(which should be ok to call from the router thread, since it takes a lock on events).
5. The device loop handles it, and does something like 
        a. `BrowserPage::FlushXR()`
        b. `mServo->RafDone(pipelineId);`
6. This does `capi::raf_done(id);`
7. Which enqueues a `WindowEvent::RafDone(PipelineID)`. 
8. Which is handled by main-thread Servo by:
        a.  Get `xr_session` from `HashMap<PipelineId, XrSession>`,
        b. `xr_session.animation_frame_request_end()`;

Note that this would significantly change the `Device` api of WebXr, and not necessarily in a worse way. The current `wait_for_animation_frame` implies that when the `SessionMsg::RequestAnimationFrame` is received the session is able to do a sync call to request a frame. This first assume the device actually wakes up the main-thread to handle the incoming msg. So it takes two events and two messages on the device to get one frame. The suggested api means only one device event and one message per frame. 

So for example `render_animation_frame` would have the device `makeCurrent` call take place in response to an event enqueued on the device loop by the router thread, which would then enqueue a `WindowEvent` in response to which "main-thread Servo" would "do the rest of the work", using `Gl`(does that require calling back into the device afterwards?). And `window.size` doesn't actually require calling into the device, as far as I can see. So one could actually make the `XRWindow` trait simpler and remove `make_current` and `swap_buffers`, since those would be made implicit to receiving the embedder msg. 

Or in the case of a main-thread WebGL call:

1. The router receives `EmbedderMsg::WebGLMsg(pipelineId, WebGLMsg)`. 
2. Get `webgl_thread, mut current_message, textures, webgl_chan` from `HashMap<PipelineId, (WebGLThread, Option<WebGLMsg>)>`, (or just `Option<WebGLThread>)`, not sure how many there are or how they are shared)
3. `current_message = Some(msg)`
3. `EventloopWaker.wake(Event::WebGL(pipelineId)`
4. `mServo->WebGL(pipelineId);`
5. `capi::webgl(pipeline_id)`
6. `WindowEvent::WegGL(PipelineID)`
7.  Get `webgl_thread, current_message, textures, webgl_chan`, and do `webgl_thread.handle_msg(current_message, textures, webgl_chan)`

So this is obviously somewhat complicated, and would require changes to the WebXr api and device code, but you'd never have to scratch your head again about what is blocking what, if anything, or why the main thread isn't waking up when you think it would, inside the currently complicated `handle_events`, which basically mixes handling events from the device, with messages from various parts of Servo. 

The embedder msg would itself be the wake-up mechanism, which enqueues an event to the device loop, which enqueues an event back on Servo, all one event at the time. 

The only question is whether `EventloopWaker.wake` is a blocking or non-blocking api. If it's blocking you might actually need a proper ipc router thread as well, since you wouldn't want to potentially block another part of the system because incoming embedder msg are not handled quickly and then have the embedder deadlock on sending a response to the previous message to that same component.

**Comment by gterzian** [comment]
More concretely in case of a Xr raf:

1. The router receives `EmbedderMsg::XRRaf(pipelineId, dest)`. 
2. It writes to a piece of data shared with the main-thread:
        a. `let timestamp = now`.
        a. Get `xr_session` from `HashMap<PipelineId, XrSession>`,
        b. `xr_session.animation_frame_request_start(timestamp, dest)`;
3. It calls into `EventloopWaker.wake(Event::Raf(pipelineId)`. 
4. This calls into a device specific way to add an event to the queue, in Hololens, something like `SendEventToServo({{Event::RAF}, {pipelineId}});`(which should be ok to call from the router thread, since it takes a lock on events).
5. The device loop handles it, and does something like 
        a. `BrowserPage::FlushXR()`
        b. `mServo->RafDone(pipelineId);`
6. This does `capi::raf_done(id);`
7. Which enqueues a `WindowEvent::RafDone(PipelineID)`. 
8. Which is handled by main-thread Servo by:
        a.  Get `xr_session` from `HashMap<PipelineId, XrSession>`,
        b. `xr_session.animation_frame_request_end()`;

Note that this would significantly change the `Device` api of WebXr, and not necessarily in a worse way. The current `wait_for_animation_frame` implies that when the `SessionMsg::RequestAnimationFrame` is received the session is able to do a sync call to request a frame. This first assume the device actually wakes up the main-thread to handle the incoming msg. So it takes two events and two messages on the device to get one frame. The suggested api means only one device event and one message per frame. 

And another example `render_animation_frame` would have the device `makeCurrent` call take place in response to an event enqueued on the device loop by the router thread, which would then enqueue a `WindowEvent` in response to which "main-thread Servo" would "do the rest of the work", using `Gl`(does that require calling back into the device afterwards?). And `window.size` doesn't actually require calling into the device, as far as I can see. So one could actually make the `XRWindow` trait simpler and remove `make_current` and `swap_buffers`, since those would be made implicit to receiving the embedder msg. 

@asajeffrey @Manishearth

**Comment by gterzian** [comment]
Or in the case of a main-thread WebGL call:

1. The router receives `EmbedderMsg::WebGLMsg(pipelineId, WebGLMsg)`. 
2. Get `webgl_thread, mut current_message, textures, webgl_chan` from `HashMap<PipelineId, (WebGLThread, Option<WebGLMsg>)>`, (or just `Option<WebGLThread>)`, not sure how many there are or how they are shared)
3. `current_message = Some(msg)`
3. `EventloopWaker.wake(Event::WebGL(pipelineId)`
4. `mServo->WebGL(pipelineId);`
5. `capi::webgl(pipeline_id)`
6. `WindowEvent::WegGL(PipelineID)`
7.  Get `webgl_thread, current_message, textures, webgl_chan`, and do `webgl_thread.handle_msg(current_message, textures, webgl_chan)`

@jdm

**Comment by gterzian** [comment]
So this is obviously somewhat complicated, and would require changes to the WebXr api and device code, but you'd never have to scratch your head again about what is blocking what, if anything, or why the main thread isn't waking up when you think it would, inside the currently complicated `handle_events`, which basically mixes handling events from the device, with messages from various parts of Servo. 

The embedder msg would itself be the wake-up mechanism, which enqueues an event to the device loop, which enqueues an event back on Servo, all one event at the time. 

The only question is whether `EventloopWaker.wake` is a blocking or non-blocking api. If it's blocking you might actually need a proper ipc router thread as well, since you wouldn't want to potentially block another part of the system because incoming embedder msg are not handled quickly and then have the embedder deadlock on sending a response to the previous message to that same component.

**Comment by gterzian** [comment]
So basically, the idea is to break up the current https://github.com/servo/servo/blob/5c8fee4e0bdf01efc83f4db10a60ac1629db6795/components/servo/lib.rs#L731

And then make all interaction with anything "running on the main-thread" become event-driven via `EmbedderMsg`. 

1. With a router thread running in the same process as the "main-thread", 
2. handling `EmbedderMsg`, with each message being a "wake-up" effectively, 
3. and then queuing events on the event-loop driving the main-thread, probably a device specific loop, through an API similar to the current "wake the event-loop" concept, but with specific events. 

Then, 
4. Each device event gives a chance for the device to actually do some necessary setup work, like flushing buffers, 
5. Then, when an actual `WindowEvent` is enqueued on the "main-thread", just added to some sort of `Vec`, then "main-thread Servo" can "know" that buffers are drained or some other setup work as already been done, meaning you can remove API's like `GlWindow.swap_buffers`, since those are implicitly done before queuing the `WindowEvent`.
6. Some API like `self.callbacks.host_callbacks.on_title_changed(title); can also dissapear, since the device loop can handle this as part of step 4 above.
7. However, the "servo keeps track of the current title" stuff can be done in response to a `WindowEvent` received after the device already did it's part. 

I can't keep myself from one more example:

When we now receive a `EmbedderMsg::AllowNavigationRequest`, we get the following sequence:

1. "servo on the main-thread" calls into an API of the "host", 
2. then creates a `WindowEvent`,
3. Then does `self.events.push(event);`
4. Finally doing `perform_updates`, which does `self.servo.handle_events(events);`, which calls into WebXr, Webgl and so on...

So that would just be:
1. Router handles `EmbedderMsg::AllowNavigationRequest`
2. Queues a `Event:AllowNavigationRequest(url)` on the device loop.
3. Queues `WindowEvent::AllowNavigationResponse(pipeline_id, data)` on "main-thread Servo"
4. End of the story, go back to blocking on a message in the router. 

To me, that makes much more sense then the current setup.

**Comment by gterzian** [comment]
Having though about it a bit more, and also looked at the Hololens and ML code, I think the problem can be identified as:

1. [`handle_events`](https://github.com/servo/servo/blob/b52bfbe68aed9b0abb39a6e994c5dad37cf7d33c/components/servo/lib.rs#L744), called via `peform_updates` is getting too complicated, as it contains many different sub-systems who all require to run on the main-thread in some cases(Compositor, WebXr, WebGL, handling of EmbedderMSg). 
2. Whenever `handle_events` is called, it's not clear why. So the code in the various event-loops contains various `try_rev` or `recv_timeout` type of logic as each sub-system cannot block, yet also doesn't quite know what to expect. 
3. The `EventLoopWaker` API is inconsistent. It's used "sometimes", mainly when a message is sent to the main-thread outside of "animation mode". During "animiation mode", it is assumed the main-thread just keeps running and therefore `peform_updates` will be regularly called. Altough in the case of "normal raf" from a document, each request does wake-up the main-thread(since it is sent to it using the compositor proxy which uses the waker on each send). 
4. A lot of those main-thread sub-systems also rely on regular "heartbeat" or "wake-ups" of the main-thread in order to be able to handle specific incoming messages. Then, as part of "handling a message" they often need to call back into the host. That seems sub-optimal, since the the call into the "host" could have been done prior to the code on main-thread Servo, as part of the initial "wake-up" of the main-thread, which could then call into a specific sub-system running on the main-thread for a specific reason, with the "device/host call" already having taken place. 


To improve the situation, we could consider to unify the "main-thread Servo" API into `embedder_traits`, used by each sub-system via the sending of `EmbedderMsg` to the main-thread, and to setup direct IPC communication between script and the embedder(currently `EmbedderMsg` are routed via the constellation. 

I also think there is an opportunity to unify Xr and "normal" raf, since I see no reason for the "normal" animation to always require routing via the constellation. Both a normal and xr raf at the end of the day are communications with the device/embedder. 

How to do it?

First, each `EmbedderMsg` would be handled by a "router" type of thread running in the same process as "main-thread Servo". 

The difference with other ipc Router threads would be that this router would always do two things:

a. If necessary, send a specific message on a crossbeam channel, to pass along some extra data that cannot be put or shouldn't be put into the `DeviceEvent`(alternatively we could pass general data via some raw pointer, but I guess it's better to just use a channel). This message would be then received with a blocking `recv` when "main-thread Servo" is called into by the device, with a specific call. 

b. Queue an `DeviceEvent` using an api similar to `EventloopWaker`, which would enqueue a specific event on the device event-loop, instead of a general "wake-up". 

c. The `DeviceEvent` would be handled on a device specific native loop, for example the current Hololens or ML code found in `support`.
 
d. The device could then do a combination of two things:
    - Do device specific work, like swapping buffers
    - Call into "main-thread Servo", essentially `peform_updates`, but for a specific event. 
    - It could interleave several of both type of calls. 

Note that whenever "main-thread Servo" is called as part of d, this would be for a specific reason. It means that the code calls could always just do one specific thing, versus running a bunch of disparate logic as part of a general `perform_updates`.  

Once the workflow would be completed, the device loop would go back to blocking on the next event. The router would also just block on an IPC channel. 

The next wake-up would then either come from an incoming `EmbedderMsg`, or a system input handled directly by the device loop. 

There would be no need for a "animation mode", since "animation mode" would simply be a result of receiving an `EmbedderMsg::RenderAF` after having sent the last frame in response to a `EmbedderMsg::RequestAF`. This means that the device loop could be simpler, and that device inputs would always be properly interleaved. 

Here is an example of a "XrRaf" workflow:

1. `XrSession.RequestAnimationFrame` is called with a callback.
2. This sends an IPC message to the embedder: `EmbedderMsg::XrRaf(FrameCallback)`
3. In the "main-process", the router thread receives `(PipelineID, EmbedderMsg::XrRaf(FrameCallback))`.
4. The router gets a crossbeam sender for that pipeline, and sends `SessionMsg::RequestAnimationFrame(callback)`.
5. The router enqueus a `DeviceEvent::XrRaf(PipelineId)`. 
6. The device loop wakes up:
    a. `FlushXR()`
    b. `mServo->XrRaf(PipelineId);`
7. "main-thread Servo" is called via `capi::xr_raf(pipeline_id);
8. `let session_thread = self.xr_sessions.get(pipeline_id)`
     - `session_thread. handle_one_msg()`.(not a while loop, just a single blocking `recv`. 
     - Here the message send at 4 is received and handled.
     - Since we know the device just swapped the buffers, handling the message is just sending a frame to script. So `GlWindowDevice` doesn't need to call into `self.window.swap_buffers()`. 

Alternatively we can keep those `WindowEvent`, and handle them in batches, I think that makes it harder to have the device "do work and then call into servo", like in the example above where buffers are swapped first, and then servo is called so that the xr session can handle the raf request. I'm also not sure what handling events in batches would bring in terms of performance, since a raf needs to always first be handled by script, which then needs to send back a message to render the animation.

**Comment by asajeffrey** [comment]
I know I'm a bit of a broken record on this subject, but this is pretty much exactly what futures 0.3 and async/await are meant to achieve: allowing different subsystems to share an API for task scheduling. It would be nice if we could take advantage of that ecosystem somehow.

**Comment by gterzian** [comment]
I'm skeptical of the value of futures or async/await for Servo. 

Why? 

1. We rarely want to "await" something, we rather want to post a task to the back of a queue and keep going. The JS code on the other end by might be "awaiting" that task, representing a pending fetch promise for example, however the Rust code does not. 
2. We usually want to keep some state in between tasks, which fits the "run a sequence of tasks on a thread, where task are received on a message-pump, and the thread can just keep some local state" model we currently have. 

Here is a concrete example:

https://github.com/servo/servo/blob/9043f247d9b031ed285e880e4b90aa523d4a63ae/components/script/fetch.rs#L132

When JS calls `Fetch`, Rust sends an IPC message to  core-resource, and sets-up an IPC route to handle incoming responses. The route then always enqueues a task on the event-loop where the fetch started. 

Note also that the Rust code then returns a promise to the JS code. 

So the Rust code isn't awaiting the fetch, it sends a message, and immediately continues by returning the promise. The JS code on the other hand can then "await" the promise, which will be resolved when the task enqueued by the router thread executes. 

The reason why it would be hard to express this workflow using async/await, is that the Rust code is implementing the runtime where the JS code is awaiting something. If the Rust code where to run somewhere in a runtime where someone else would give you async APIs to use, it would indeed be a good fit. Say a server using an async I/O API. As a matter of fact, we use futures inside of fetch, and there they are a good fit. 

https://github.com/servo/servo/blob/73784246b84295e5775d9d2e6d5fc7a21a5bdab8/components/net/http_loader.rs#L1354

---- 

On a more general note, async/await is essentially an event-driven style(with or without a threadpool, with a threadpool perhaps making things even more complicated), and the problem with using such a style as a general approach to concurrency/parallelism as been noted by others in the past:

> The “threads vs. events” question is misphrased. 
> Locks and top-level event loops are both evil.
> Threads with events (message passing) works very well. 
> https://swtch.com/~rsc/talks/threads07/#(38)

> A further problem with the event-driven style is that the resulting code becomes quite difficult to understand, maintain, and debug. This is primarily because it is harder to tell from reading the code which routine caused the current one to be called, and which routine will be called next. Standard debugging and performance tools become less effective too, as they rarely have support for tracing through continuations, and continuation mechanisms rarely maintain call history. In contrast, non-event-driven programs maintain much of their state and history conveniently on the thread stacks; debugging tools can reveal a great deal simply by displaying stack traces. https://www.chromium.org/developers/lock-and-condition-variable

> There are a number of ways to handle concurrency, and ways to avoid it altogether. Of the various possible models, only two permit high-performance implementations that can use multiple CPUs and sharing of resources, and still allow large programs to be built from smaller components while maintaining abstraction boundaries. These models are "threads + mutexes + condition-variables", and "threads + message-passing". These two can be used together, and often are. https://www.chromium.org/developers/lock-and-condition-variable 

----

Finally, we could look at other web engines out there, see if they use some sort of event-driven style. 

The threading model in Chromium is almost entirely based on message-passing, where their `PostTask` api is eerily similar to our own `task!` macro. 

The different is mainly that they run "Sequences" of tasks on a "virtual thread", with a scheduler posting the task on potentially different physical threads, but retaining their sequential ordering.

We do this by simply having a thread with a message-pump. 

Note that their scheduler also bring problems:

1. It's harder to do microtasks checkpoints: 

>  The main objective is to fix the unit of microtask queues which is currently wrong in Blink (July 5, 2019) https://docs.google.com/document/d/1GSHp-KoP4M1oq-vQaw8W8g93lm_k1zsykP_GAzRrbYo/edit

2. For some features, they have to move away from the scheduler and into something that will make you smile:

> Proposed Change for the spec aligned Event Loop
> Replaces per-frame scheduler with per eTLD+1 Event Loop.
> Two browsing contexts can be the same origin-domain browsing contexts if and only if they have the same eTLD+1 domain. As it’s hard to merge or unmerge task queues, per eTLD+1 is the easiest grain size to implement the spec compliant execution model. https://docs.google.com/document/d/1bM_1SjE25NMQrD4IJFAPc1tI4i8tvSET2f0IgjukRho/ 

----

So my opinion is that:

1. async/await isn't likely to be a good fit for systems implementing async runtimes for other code to run. It might even not be a good fit for general concurrency beyond some specific async i/o use cases. 
2. Native threads a pretty good, more "sophisticated schedulers" can make it harder to implement a spec compliant execution model. Our "script-thread" is basically a scheduler for tasks from all document sharing an event-loop. We could try to use a scheduler for say, dedicated workers, and it would be hard(since it's not Rust code you want to yield, it's the [JS task you want to cooperatively schedule](https://docs.google.com/document/d/1bM_1SjE25NMQrD4IJFAPc1tI4i8tvSET2f0IgjukRho/edit#) to ensure minimal latency, and could bring unexpected adverse consequences for correctness.

**Comment by asajeffrey** [comment]
You're raising an interesting, but larger, point which is whether async/await is a good fit for Servo in general, and I see your skepticism. The event loop model that HTML requires is complex, and it is not obvious whether it can be fitted into the one-size-fits-all futures model. It would require at least a custom executor, that followed the semantics of event loops, microtasks, etc.

For the main thread, the places we're blocking is (off the top of my head):
- for windowing events (when there are no animations or XR running)
- for window vsync in `swap_buffers` (when there are animations running)
- for the next XR animation frame (when there's an active XR session on the main thread)
- when webrender requests an external image (complicated by WebGL possibly running on the main thread)

At the moment, we have nasty ad hoc logic to deal with these various forms of blocking, e.g. in XR waiting 5ms for the animation frame to process. Really we'd like a mechanism for saving the current state, and moving on to other tasks, where the state will be restored when appropriate. This is pretty much exactly what async/await does, with the advantage of having compiler support so the state machine is generated by rustc rather than by hand.

And the answer to your question "wouldn't you be facing the additional challenge of integrating a device-specific main-loop with async Rust?" is yes, we probably would need each embedding to implement a custom executor, e.g. the one for glutin would need to use glutin's event loop waker.

**Comment by gterzian** [comment]
I agree async/await could be a good fit for the the main-thread logic, and it could be easier to first do it by hand? 

I can imagine the device loop running events based on the wake-up, and then calling into Servo, see for a concrete example: https://github.com/servo/servo/issues/23823#issuecomment-516348324

And if we re-shuffle some of the  webgl and webxr API, I think we could make the entire thing "non-blocking"(unless there is nothing to do). 

For example: 

1. `EmbedderMsg::XrRaf(pipelineID)`.
2. `EventloopWaker::wake(DeviceEvent::XrRaf(id))`
3. Swap the buffers first, then call into main-thread Servo:
```c++
case Event::XrRaf:
        auto [id] = e.pipelineId;
        Flush();
        mServo->XrRaf(id);
        break;
```
4. In Servo, no need to call `window.swap_buffers`, since that's already done, instead the frame is immediately sent to script.

**Comment by asajeffrey** [comment]
Should we move this discussion over to https://github.com/servo/servo/issues/23899?

**Comment by asajeffrey** [comment]
(by which I mean the general discussion about how to tame the main thread, not the particular discussion about XR rAF messages, which should probably stay here.)

**Comment by asajeffrey** [comment]
@gterzian the problem we're hitting with XR rAFs and the main embedder event loop is partly caused by there being quite different use cases. Some are devices like hololens or magicleap, where the XR device is very tightly wedded to the embedder, and some are tethered devices like HTC Vive or Oculus Rift where the browser is the stock browser for the host PC's OS, and the embedder knows little about the XR device driver. We're trying to support both of these use-cases with one API, which is part of why we're running into problems.

In particular, for tethered VR headsets, there's two APIs for "wait for the next frame", the VR device's one and the host OS's one, and they operate at different cadences, sigh, and only offer blocking APIs for waiting, double sigh, and some devices have main thread affinity.

**Comment by EmmaParker123** [comment]
Nice Article,

Is your PST file corrupt? Looking for a solution to fix and recover damaged and corrupt files? If you are getting these issues then 

we suggest you download the **[Microsoft Scanpst](http://www.microsoftscanpst.com)** Software.
